module/zfs/dmu_recv.c

```diff
@@ -2043,6 +2275,60 @@ receive_process_write_record(struct receive_writer_arg *rwa,
 	    !DMU_OT_IS_VALID(drrw->drr_type))
 		return (SET_ERROR(EINVAL));
 
+	if (rwa->heal) {
+		blkptr_t *bp;
+		dmu_buf_t *dbp;
+		dnode_t *dn;
+		uint8_t buf[1];
+		int flags = DMU_READ_PREFETCH;
+
+		if (rwa->raw)
+			flags |= DMU_READ_NO_DECRYPT;
+
+		if (rwa->byteswap) {
+			dmu_object_byteswap_t byteswap =
+			    DMU_OT_BYTESWAP(drrw->drr_type);
+			dmu_ot_byteswap[byteswap].ob_func(abd_to_buf(rrd->abd),
+			    DRR_WRITE_PAYLOAD_SIZE(drrw));
+		}
+
+		err = dmu_buf_hold_noread(rwa->os, drrw->drr_object,
+		    drrw->drr_offset, FTAG, &dbp);
+		if (err != 0)
+			return (err);
+
+		/* Try to read the object to see if it needs healing */
+		err = dmu_read(rwa->os, drrw->drr_object, drrw->drr_offset,
```

Comment:
This can be dbuf_read which is a little cleaner.
```
diff --git a/module/zfs/dmu_recv.c b/module/zfs/dmu_recv.c
index 5d7c9ad88..595e01df2 100644
--- a/module/zfs/dmu_recv.c
+++ b/module/zfs/dmu_recv.c
@@ -2279,7 +2279,6 @@ receive_process_write_record(struct receive_writer_arg *rwa,
                blkptr_t *bp;
                dmu_buf_t *dbp;
                dnode_t *dn;
-               uint8_t buf[1];
                int flags = DMU_READ_PREFETCH;
 
                if (rwa->raw)
@@ -2298,10 +2297,10 @@ receive_process_write_record(struct receive_writer_arg *rwa,
                        return (err);
 
                /* Try to read the object to see if it needs healing */
-               err = dmu_read(rwa->os, drrw->drr_object, drrw->drr_offset,
-                   sizeof (buf), buf, flags);
+               err = dbuf_read((dmu_buf_impl_t *)dbp, NULL, DB_RF_CANFAIL);
+
                /*
-                * We only try to heal when dmu_read() returns a ECKSUMs.
+                * We only try to heal when dbuf_read() returns ECKSUM.
                 * Other errors (even EIO) get returned to caller.
                 * EIO indicates that the device is not present/accessible,
                 * so writing to it will likely fail.
```